README.rst
monthday
========

.. image:: https://badge.fury.io/py/monthday.svg
   :alt: PyPI
   :target: https://pypi.python.org/pypi/monthday

.. image:: https://readthedocs.org/projects/monthday/badge/
   :alt: Read the Docs
   :target: https://monthday.readthedocs.org/

.. image:: https://travis-ci.org/dahlia/monthday.svg?branch=master
   :alt: Build Status
   :target: https://travis-ci.org/dahlia/monthday

.. image:: https://codecov.io/github/dahlia/monthday/coverage.svg?branch=master
   :alt: Coverage Status
   :target: https://codecov.io/github/dahlia/monthday?branch=master

This package provides the ``MonthDay`` value type for dealing dates without
year.  It is useful for dealing with birthdays, or anniversaries.
Works on Python 2.6, 2.7, 3.2--3.5, PyPy, PyPy3.

.. code-block:: pycon

   >>> from monthday import *
   >>> aug_4 = MonthDay(8, 4)
   >>> aug_4
   monthday.MonthDay(8, 4)
   >>> aug_4.date(1988)
   datetime.date(1988, 8, 4)
   >>> list(aug_4.dates(range(2013, 2016)))
   [datetime.date(2013, 8, 4),
    datetime.date(2014, 8, 4),
    datetime.date(2015, 8, 4)]
   >>> from datetime import date
   >>> MonthDay.from_date(date(2015, 12, 25))
   monthday.MonthDay(12, 25)

It's available on PyPI__:

.. code-block:: console

   $ pip install monthday

Written by `Hong Minhee`__, and distributed under LGPLv3_ or later.
Find the source code from the `GitHub repository`__.

__ https://pypi.python.org/pypi/monthday
__ http://hongminhee.org/
.. _LGPLv3: http://www.gnu.org/licenses/lgpl-3.0.html
__ https://github.com/dahlia/monthday


CHANGES.rst
Changelog
=========

Version 0.9.1
-------------

To be released.


Version 0.9.0
-------------

The first alpha version.  Released on November 29, 2015.


monthday_test.py
import datetime
import pickle

from pytest import fixture, raises

from monthday import MonthDay


@fixture
def feb_29():
    """February 29.  Only available for leap years."""
    return MonthDay(2, 29)


@fixture
def aug_4():
    """August 4.  Author's birthday."""
    return MonthDay(8, 4)


@fixture
def dec_25():
    """December 25.  Christmas."""
    return MonthDay(12, 25)


def test_month_day_from_date(aug_4, dec_25):
    assert MonthDay.from_date(datetime.date(1988, 8, 4)) == aug_4
    assert MonthDay.from_date(datetime.date(2015, 12, 25)) == dec_25
    assert MonthDay.from_date(datetime.datetime(1988, 8, 4)) == aug_4


def test_month_day_from_date_type_error():
    with raises(TypeError):
        MonthDay.from_date('1988-08-04')
    with raises(TypeError):
        MonthDay.from_date(19880804)


def test_month_day_init():
    aug_4 = MonthDay(8, 4)
    assert aug_4.month == 8
    assert aug_4.day == 4


def test_month_day_init_value_error():
    # Invalid months
    with raises(ValueError):
        MonthDay(-12, 1)
    with raises(ValueError):
        MonthDay(-1, 1)
    with raises(ValueError):
        MonthDay(0, 1)
    # Invalid days
    with raises(ValueError):
        MonthDay(1, -31)
    with raises(ValueError):
        MonthDay(1, -1)
    with raises(ValueError):
        MonthDay(1, 0)
    with raises(ValueError):
        MonthDay(1, 32)
    with raises(ValueError):
        MonthDay(2, 30)
    with raises(ValueError):
        MonthDay(2, 31)
    with raises(ValueError):
        MonthDay(3, 32)
    with raises(ValueError):
        MonthDay(4, 31)
    with raises(ValueError):
        MonthDay(5, 32)
    with raises(ValueError):
        MonthDay(6, 31)
    with raises(ValueError):
        MonthDay(7, 32)
    with raises(ValueError):
        MonthDay(8, 32)
    with raises(ValueError):
        MonthDay(9, 31)
    with raises(ValueError):
        MonthDay(10, 32)
    with raises(ValueError):
        MonthDay(11, 31)
    with raises(ValueError):
        MonthDay(12, 32)


def test_month_day_init_type_error():
    with raises(TypeError):
        MonthDay('8', 4)
    with raises(TypeError):
        MonthDay(8.0, 4)
    with raises(TypeError):
        MonthDay(8, '4')
    with raises(TypeError):
        MonthDay(8, 4.0)


def test_month_day_equality(aug_4, dec_25):
    assert aug_4 == MonthDay(8, 4)
    assert not aug_4 != MonthDay(8, 4)
    assert hash(aug_4) == hash(MonthDay(8, 4))
    assert dec_25 == MonthDay(12, 25)
    assert not dec_25 != MonthDay(12, 25)
    assert hash(dec_25) == hash(MonthDay(12, 25))
    assert aug_4 != MonthDay(8, 5)
    assert not aug_4 == MonthDay(8, 5)
    assert hash(aug_4) != hash(MonthDay(8, 5))
    assert aug_4 != dec_25
    assert not aug_4 == dec_25
    assert hash(aug_4) != hash(dec_25)
    assert dec_25 != MonthDay(12, 24)
    assert not dec_25 == MonthDay(12, 24)
    assert hash(dec_25) != hash(MonthDay(12, 24))
    assert dec_25 != aug_4
    assert not dec_25 == aug_4
    assert hash(dec_25) != hash(aug_4)


def test_month_day_date(aug_4, feb_29):
    assert aug_4.date(1988) == datetime.date(1988, 8, 4)
    assert feb_29.date(2016) == datetime.date(2016, 2, 29)
    with raises(ValueError) as excinfo:
        feb_29.date(2015)
    assert str(excinfo.value) == '''since 2015 is not a leap year, monthday.\
MonthDay(2, 29) can't be combined with 2015'''
    with raises(ValueError) as excinfo:
        feb_29.date(2014)
    with raises(ValueError):
        feb_29.date(2013)
    assert feb_29.date(2012) == datetime.date(2012, 2, 29)
    with raises(ValueError):
        feb_29.date(2011)
    with raises(TypeError):
        aug_4.date('1988')


def test_month_day_dates(aug_4, feb_29):
    assert list(aug_4.dates([])) == []
    assert list(aug_4.dates(range(1988, 1992))) == [
        datetime.date(1988, 8, 4),
        datetime.date(1989, 8, 4),
        datetime.date(1990, 8, 4),
        datetime.date(1991, 8, 4),
    ]
    years = range(2011, 2017)
    with raises(ValueError):
        list(feb_29.dates(years))
    with raises(ValueError):
        list(feb_29.dates(years, error_invalid_dates=True))
    assert list(feb_29.dates(years, error_invalid_dates=False)) == [
        datetime.date(2012, 2, 29), datetime.date(2016, 2, 29),
    ]
    assert list(feb_29.dates(years, error_invalid_dates=None)) == [
        None, datetime.date(2012, 2, 29),
        None, None, None, datetime.date(2016, 2, 29),
    ]
    with raises(TypeError):
        aug_4.dates(1988)


def test_month_day_str(aug_4, dec_25):
    assert str(aug_4) == '08-04'
    assert str(dec_25) == '12-25'


def test_month_day_repr(aug_4, dec_25):
    assert repr(aug_4) == 'monthday.MonthDay(8, 4)'
    assert repr(dec_25) == 'monthday.MonthDay(12, 25)'


def test_month_day_pickle(aug_4, dec_25):
    assert pickle.loads(pickle.dumps(aug_4)) == aug_4
    assert pickle.loads(pickle.dumps(dec_25)) == dec_25


monthday.py
""":mod:`monthday` --- Date without year
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

"""
import collections
import datetime
import numbers


__all__ = 'MonthDay', '__version__'
__version__ = '0.9.1'


class MonthDay(object):
    """Date without year.  Useful for birthdays, or anniversaries.

    :param month: a month number, from 1 to 12
    :type month: :class:`numbers.Integral`
    :param day: a day of the ``month``, from 1 to 31 (or 30, or 29)
    :type day: :class:`numbers.Integral`
    :raise ValueError: if ``month`` or ``date`` is out of valid range

    .. attribute:: month

       (:class:`numbers.Integral`) The month number, from 1 to 12.

    .. attribute:: day

       (:class:`numbers.Integral`) The day of the ``month``, from 1 to 31.

    """

    __slots__ = 'month', 'day'

    @classmethod
    def from_date(cls, date):
        """Get only :class:`MonthDay` from the given ``date``.

        :param date: the date or date/time
        :type date: :class:`datetime.date`, :class:`datetime.datetime`
        :return: :class:`MonthDay` without ``date``'s
                 :attr:`~datetime.date.year`
        :rtype: :class:`MonthDay`

        """
        if isinstance(date, datetime.date):
            return cls(date.month, date.day)
        raise TypeError('date must be an instance of datetime.date, not ' +
                        repr(date))

    def __init__(self, month, day):
        if not isinstance(month, numbers.Integral):
            raise TypeError('month must be an integer, not ' + repr(month))
        elif not isinstance(day, numbers.Integral):
            raise TypeError('day must be an integer, not ' + repr(day))
        elif not 1 <= month <= 12:
            raise ValueError('month must be from 1 to 12, not ' + repr(month))
        elif month in (1, 3, 5, 7, 8, 10, 12) and not 1 <= day <= 31:
            raise ValueError('day must be from 1 to 31 for month={!r}, but '
                             '{!r} was given'.format(month, day))
        elif month in (2, 4, 6, 9, 11) and not 1 <= day <= 30:
            raise ValueError('day must be from 1 to 30 for month={!r}, but '
                             '{!r} was given'.format(month, day))
        elif month == 2 and not 1 <= day <= 29:
            raise ValueError('day must be from 1 to 29 for month=2, but '
                             '{!r} was given'.format(day))
        self.month = int(month)
        self.day = int(day)

    def __eq__(self, other):
        return (isinstance(other, type(self)) and
                self.month == other.month and
                self.day == other.day)

    def __ne__(self, other):
        return not self == other

    def __hash__(self):
        return self.month * 100 + self.day

    def date(self, year):
        """Get a :class:`~datetime.date` by combining the given ``year``
        with it.

        >>> MonthDay(12, 25).date(2015)
        datetime.date(2015, 12, 25)

        It may raise :exc:`ValueError` if February 29 is tried to be combined
        with a non-leap year e.g.:

        >>> feb_29 = MonthDay(2, 29)
        >>> feb_29.date(2012)
        datetime.date(2012, 2, 29)
        >>> feb_29.date(2013)
        Traceback (most recent call last):
          ...
        ValueError: since 2013 is not a leap year,
                    monthday.MonthDay(2, 29) can't be combined with 2013

        :param year: a year to combine with
        :type year: :class:`numbers.Integral`
        :return: a :class:`datetime.date` with the given ``year``
        :rtype: :class:`datetime.date`
        :raise ValueError: when ``year`` is not a leap year
                           while it's ``MonthDay(2, 29)``

        """
        if not isinstance(year, numbers.Integral):
            raise TypeError('year must be an integer, not ' + repr(year))
        try:
            return datetime.date(int(year), self.month, self.day)
        except ValueError:
            if self.month == 2 and self.day == 29:
                raise ValueError("since {0!r} is not a leap year, {1!r} can't "
                                 "be combined with {0!r}".format(year, self))
            raise

    def dates(self, years, error_invalid_dates=True):
        """Get :class:`~datetime.date`\ s by combining the given ``years``
        with it.

        >>> list(MonthDay(8, 4).dates(range(1988, 1992)))
        [datetime.date(1988, 8, 4), datetime.date(1989, 8, 4),
         datetime.date(1990, 8, 4), datetime.date(1991, 8, 4)]

        It may raise :exc:`ValueError` if there happen to be any invalid
        dates in the result, e.g. Feburary 29 for non-leap years:

        >>> feb_29 = MonthDay(2, 29)
        >>> list(feb_29.dates(range(2011, 2017)))
        Traceback (most recent call last):
          ...
        ValueError: since 2010 is not a leap year,
                    monthday.MonthDay(2, 29) can't be combined with 2010

        If you want to simply ignore these invalid dates in the result,
        set ``error_invalid_dates`` to :const:`False` e.g.:

        >>> list(feb_29.dates(range(2011, 2017), error_invalid_dates=False))
        [datetime.date(2012, 2, 29), datetime.date(2016, 2, 29)]

        But the result length might be shorter than the input ``years`` list.

        If you want to match the length of the input and the result, set
        ``error_invalid_dates`` to :const:`None` --- it will replace invalid
        dates in the result with :const:`None` values e.g.:

        >>> list(feb_29.dates(range(2011, 2017), error_invalid_dates=None))
        [None, datetime.date(2012, 2, 29),
         None, None, None, datetime.date(2016, 2, 29)]

        :param years: years to combine with
        :type years: :class:`~collections.abc.Iterable`
        :param error_invalid_dates: if set to :const:`True`, raise
                                    :exc:`ValueError` for invalid dates.
                                    if set to :const:`False`, just ignore
                                    invalid dates --- the result length might
                                    be shorter than the input ``years`` list.
                                    if set to :const:`None`, fill :const:`None`
                                    values instead of invalid dates --- the
                                    result length must be the same to the input
                                    ``year`` list.  :const:`True` by default
        :type error_invalid_dates: :class:`bool`, ``type(None)``
        :return: :class:`datetime.date` values with the given ``years``.
                 the order corresponds to the input ``years``' order
        :rtype: :class:`~collections.abc.Iterable`
        :raise ValueError: if ``error_invalid_dates`` is set to :const:`True`
                           and there happend to be any invalid dates in
                           the result
        :raise TypeError: if ``years`` is not iterable of integers

        """
        if not isinstance(years, collections.Iterable):
            raise TypeError('years must be iterable, not ' + repr(years))

        def generate():
            for year in years:
                try:
                    yield self.date(year)
                except ValueError:
                    if error_invalid_dates:
                        raise
                    elif error_invalid_dates is None:
                        yield None
        return generate()

    def __getstate__(self):
        return self.month, self.day

    def __setstate__(self, state):
        self.month, self.day = state

    def __str__(self):
        return '{0:02d}-{1:02d}'.format(self.month, self.day)

    def __repr__(self):
        return '{0.__module__}.{0.__name__}({1!r}, {2!r})'.format(
            type(self), self.month, self.day
        )
